### ISSUE TITLE
WGSL: Support the `pointer_composite_access` language extension

### ISSUE THREAD

**Issue by rcoreilly** [gfx-rs/wgpu#6192]
**Description**

The WGSL spec has a `requires` directive to request language extensions, such as `pointer_composite_access` https://www.w3.org/TR/WGSL/#language-extensions-sec

When I put that at the top of my .wgsl shader, it barfs:

```
[wgpu] [Error] Device::create_shader_module error: 
Shader 'squares' parsing error: expected global item ('struct', 'const', 'var', 'alias', ';', 'fn') or the end of the file, found 'requires'
  ┌─ wgsl:2:1
  │
2 │ requires pointer_composite_access;
  │ ^^^^^^^^ expected global item ('struct', 'const', 'var', 'alias', ';', 'fn') or the end of the file
```

I would particularly be interested in when that specific language extension `pointer_composite_access` will be included as well -- huge improvement in usability.  Also `unrestricted_pointer_parameters` seems rather important if I understand it properly, which I'm not sure I do..

**Repro steps**
add that line to the top of any .wgsl code.

**Expected vs observed behavior**

didn't work.

**Platform**

Mac OS, latest trunk HEAD 26398ea using `naga` cli, and [v0.19.4.1](https://github.com/gfx-rs/wgpu-native/releases/tag/v0.19.4.1) of wgpu-native.

**Comment by ErichDonGubler** [comment]
@rcoreilly: Basic support for parsing `requires` is now posted in #6437. As the scope of parsing is in https://github.com/gfx-rs/wgpu/issues/6350, I'll narrow the scope of this issue (and edit the OP) to just the `pointer_composite_access` extension.

**Comment by sagudev** [comment]
I am working on this.

**Comment by ErichDonGubler** [comment]
@sagudev: Sweet! I'll plan on reviewing your PRs. It might be next year when I have bandwidth to do so, but I think it's the same for all the Firefox WebGPU/Naga team. 😅

**Comment by sagudev** [comment]
It would be nice to get https://github.com/gfx-rs/wgpu/pull/6814 landed before (holidays).

**Comment by sagudev** [comment]
I think we need to remove all usages of `Error::Pointer`, ex: https://github.com/gfx-rs/wgpu/blob/ee3ae0e549fe01c4b699cf68f9b67ae8ea807564/naga/src/front/wgsl/lower/mod.rs#L1981

Also I think some types of expressions are already supported, so It would be really interesting to see CTS results with pointer_composite_access reported as supported but otherwise unimpl.

**Comment by sagudev** [comment]
> Also I think some types of expressions are already supported, so It would be really interesting to see CTS results with pointer_composite_access reported as supported but otherwise unimpl.

Eh, no: https://github.com/sagudev/servo/actions/runs/12721814335/attempts/1#summary-35466365144 (except on some subtests).

**Comment by newDINO** [comment]
Is this done? I'm using naga 24.0.0 and it doesn't seem to support it:
`error: the 'pointer_composite_access' language extension is not yet supported`

This feature could be quite important to OOP style.
But I can currently use the following workaround
```wgsl
fn g(data_ptr: ptr<private, Data>) {
    var data = *data_ptr;
    // make some changes to data
    *data_ptr = data;
}
```
I don't know whether this would be slower than direct pointer_composite_access.

**Comment by cwfitzgerald** [comment]
Yes, the PR landed a week after v24 was cut, so will appear in v25 in about a month
